File: oa-portal/problems/m24-tree-of-space/starters/main.java
import java.io.*;
import java.util.*;

public class Main {

    /* parent[v] is the parent of node v; parent[1] is 1 (the root).
       ops[i] = {type, node, uid}, type 1 lock / 2 unlock / 3 upgrade. */
    static boolean[] runOperations(int n, int[] parent, int[][] ops) {
        // write your code here

        return new boolean[ops.length];
    }

    public static void main(String[] args) throws IOException {
        FastReader in = new FastReader();
        int n = in.nextInt(), q = in.nextInt();
        int[] parent = new int[n + 1];
        parent[1] = 1;
        for (int i = 2; i <= n; i++) parent[i] = in.nextInt();
        int[][] ops = new int[q][3];
        for (int i = 0; i < q; i++) {
            ops[i][0] = in.nextInt(); ops[i][1] = in.nextInt(); ops[i][2] = in.nextInt();
        }

        StringBuilder sb = new StringBuilder();
        for (boolean ok : runOperations(n, parent, ops)) sb.append(ok ? "true" : "false").append('\n');
        System.out.print(sb);
    }

    static class FastReader {
        private final DataInputStream in = new DataInputStream(new BufferedInputStream(System.in, 1 << 16));
        int nextInt() throws IOException { return (int) nextLong(); }
        long nextLong() throws IOException {
            int b = in.read();
            while (b != '-' && (b < '0' || b > '9')) b = in.read();
            boolean neg = b == '-';
            if (neg) b = in.read();
            long v = 0;
            while (b >= '0' && b <= '9') { v = v * 10 + (b - '0'); b = in.read(); }
            return neg ? -v : v;
        }
        String next() throws IOException {
            int b = in.read();
            while (b <= ' ') b = in.read();
            StringBuilder sb = new StringBuilder();
            while (b > ' ') { sb.append((char) b); b = in.read(); }
            return sb.toString();
        }
    }
}

File: oa-portal/problems/m24-tree-of-space/starters/main.cpp
#include <bits/stdc++.h>
using namespace std;

/* parent[v] is the parent of node v; parent[1] is 1 (the root).
   ops[i] = {type, node, uid}, type 1 lock / 2 unlock / 3 upgrade.
   Return one result per operation, in order. */
vector<bool> runOperations(int n, const vector<int>& parent, const vector<array<int,3>>& ops) {
    // write your code here

    return vector<bool>(ops.size(), false);
}

int main() {
    int n, q;
    if (scanf("%d %d", &n, &q) != 2) return 0;
    vector<int> parent(n + 1, 1);
    for (int i = 2; i <= n; i++) scanf("%d", &parent[i]);
    vector<array<int,3>> ops(q);
    for (auto &o : ops) scanf("%d %d %d", &o[0], &o[1], &o[2]);

    string out;
    for (bool ok : runOperations(n, parent, ops)) out += ok ? "true\n" : "false\n";
    fwrite(out.data(), 1, out.size(), stdout);
    return 0;
}
